### ISSUE TITLE
enslaved NIC failed to rename after lost/gained carrier, due to "Device or resource busy"

### ISSUE THREAD

**Issue by enihcam** [systemd/systemd#20895]
**systemd version the issue has been seen with**
 > systemd 249 (249.4-1-arch)
+PAM +AUDIT -SELINUX -APPARMOR -IMA +SMACK +SECCOMP +GCRYPT +GNUTLS +OPENSSL +ACL +BLKID +CURL +ELFUTILS +FIDO2 +IDN2 -IDN +IPTC +KMOD +LIBCRYPTSETUP +LIBFDISK +PCRE2 -PWQUALITY +P11KIT -QRENCODE +BZIP2 +LZ4 +XZ +ZLIB +ZSTD +XKBCOMMON +UTMP -SYSVINIT default-hierarchy=unified

**Used distribution**
 > Archlinux

**Linux kernel version used** (`uname -a`)

 > `Linux 5.10.69-xanmod1-1-lts #2 SMP PREEMPT Mon, 27 Sep 2021 00:25:44 +0000 x86_64 GNU/Linux`

**CPU architecture issue was seen on**
 > Intel(R) Xeon(R) Platinum 8272CL CPU @ 2.60GHz

**Expected behaviour you didn't see**
 > enslaved NIC should be renamed from 'eth1' to 'enP33443s1' after lost/gained carrier

**Unexpected behaviour you saw**
 > Sep 30 08:43:29 systemd-networkd[246]: enP33443s1: Link DOWN
Sep 30 08:43:29 systemd-networkd[246]: enP33443s1: Lost carrier
Sep 30 08:45:49 systemd-networkd[246]: eth1: Link UP
Sep 30 08:45:49 systemd-networkd[246]: eth1: Gained carrier
Sep 30 08:45:49 systemd-udevd[49841]: eth1: Failed to rename network interface 4 from 'eth1' to 'enP33443s1': Device or resource busy

**Steps to reproduce the problem**
 > unplug/plug cable i guess?

**Additional program output to the terminal or log subsystem illustrating the issue**

```text
$ lspci
0000:00:00.0 Host bridge: Intel Corporation 440BX/ZX/DX - 82443BX/ZX/DX Host bridge (AGP disabled) (rev 03)
0000:00:07.0 ISA bridge: Intel Corporation 82371AB/EB/MB PIIX4 ISA (rev 01)
0000:00:07.1 IDE interface: Intel Corporation 82371AB/EB/MB PIIX4 IDE (rev 01)
0000:00:07.3 Bridge: Intel Corporation 82371AB/EB/MB PIIX4 ACPI (rev 02)
0000:00:08.0 VGA compatible controller: Microsoft Corporation Hyper-V virtual VGA
82a3:00:02.0 Ethernet controller: Mellanox Technologies MT27710 Family [ConnectX-4 Lx Virtual Function] (rev 80)

$ networkctl status eth1
● 4: eth1
             Link File: n/a
          Network File: n/a
                  Type: ether
                 State: enslaved (pending)
          Online state: unknown
            HW Address: 00:0d:3a:xx:yy:zz (Microsoft Corp.)
                   MTU: 1500 (min: 68, max: 9978)
                 QDisc: mq
                Master: eth0
  Queue Length (Tx/Rx): 16/4
      Auto negotiation: no
                 Speed: 50Gbps
     Activation Policy: up
   Required For Online: yes

Sep 27 08:35:08 systemd-networkd[246]: eth1: Interface name change detected, renamed to enP33443s1.
Sep 30 08:45:49 systemd-networkd[246]: eth1: Link UP
Sep 30 08:45:49 systemd-networkd[246]: eth1: Gained carrier
Sep 30 08:45:49 systemd-udevd[49841]: eth1: Failed to rename network interface 4 from 'eth1' to 'enP33443s1': Device or resource busy

$ cat 10-default.link
[Match]
Driver=mlx5*

[Link]
NamePolicy=keep kernel database onboard slot path
MACAddressPolicy=persistent
RxBufferSize=8192
TxBufferSize=8192

$ cat 20-default.network
[Match]
Driver=hv_netvsc

[Link]
ActivationPolicy=always-up

[Network]
Address=10.0.0.4/24
ConfigureWithoutCarrier=true
DHCP=no
DNS=127.0.0.1
Gateway=10.0.0.1
IgnoreCarrierLoss=true
IPv6AcceptRA=no
LinkLocalAddressing=no
LLDP=false
LLMNR=false

[Route]
FastOpenNoCookie=true
Gateway=10.0.0.1
InitialAdvertisedReceiveWindow=88
InitialCongestionWindow=88
QuickAck=true
```

**Comment by yuwata** [comment]
Hmm, It seems this is related to #20657 (but not a duplicate). Why does networkd once start to configure the interface?? Arch linux has some special udev rules or .link files??

**Comment by yuwata** [comment]
I mean, networkd starts to configure interfaces when they are once processed by udevd. At least currently, networkd does not expect that, after networkd started to manage interfaces, some other tools (including udevd) change configs e.g. interface name. Does udevd process network interfaces twice for some reason??

**Comment by yuwata** [comment]
Please try to boot with debugging option enabled on networkd and udevd, and provide the logs (`journalctl -b --no-hostname -u systemd-networkd.service -u systemd-udevd.service`). The debugging logs can be generated by creating the following two files.
```
# /etc/systemd/system/systemd-udevd.service.d/override.conf
[Service]
Environment=SYSTEMD_LOG_LEVEL=debug
LogRateLimitIntervalSec=0
```
```
# /etc/systemd/system/systemd-networkd.service.d/override.conf
[Service]
Environment=SYSTEMD_LOG_LEVEL=debug
```

**Comment by yuwata** [comment]
I notices #20657 is on gentoo. So, it seems such issues do not depend on distribution.

**Comment by yuwata** [comment]
Hm? Interface name is conflicted? Please show the result of `ip link`.

**Comment by yuwata** [comment]
I am very confused. Why unplugging and plugging cable triggers uevents?

**Comment by enihcam** [comment]
> I am very confused. Why unplugging and plugging cable triggers uevents?

actually I didn't know why the carrier was lost, so i was just guessing. the machine is a virtual machine being hosted on Azure, which means i am unable to reproduce the issue and show 'ip link'. sorry.

**Comment by yuwata** [comment]
I hope this is already fixed by 6681eb021a0b56ef0dc849e3b358a515ece16482. Closing.

If the commit does not fix this issue. Feel free to request to reopen this page. But please provide debugging logs of networkd and udevd.
